test-data/unit/check-generics.test

```diff
@@ -1814,3 +1814,98 @@ def g(x: T) -> T: return x
+T = TypeVar('T')
+class A(Generic[T]):
+    @classmethod
+    def f(cls) -> T: ...
+class B(A[str]): pass
+class C(A[int]): pass
+reveal_type(B.f)  # E: Revealed type is 'def () -> builtins.str*'
+reveal_type(B.f())  # E: Revealed type is 'builtins.str*'
+reveal_type(C.f)  # E: Revealed type is 'def () -> builtins.int*'
+reveal_type(C.f())  # E: Revealed type is 'builtins.int*'
+[builtins fixtures/classmethod.pyi]
+[out]
+
+[case testClassMethodOfGenericClassMultipleInheritance]
+from typing import TypeVar, Generic
+t = TypeVar('t')
+class A(Generic[t]):
+    @classmethod
+    def f(cls) -> t: ...
+class B(A[str]): pass
+class C(A[int]): pass
+reveal_type(B.f)  # E: Revealed type is 'def () -> builtins.str*'
+reveal_type(B.f())  # E: Revealed type is 'builtins.str*'
+reveal_type(C.f)  # E: Revealed type is 'def () -> builtins.int*'
+reveal_type(C.f())  # E: Revealed type is 'builtins.int*'
+[builtins fixtures/classmethod.pyi]
+[out]
+
+[case testClassMethodOfGenericClassMultipleLevelInheritance]
+from typing import TypeVar, Generic
+t = TypeVar('t')
+class A(Generic[t]):
+    @classmethod
+    def f(cls) -> t: ...
+class B(A[str]): pass
+class C(B): pass
+class D(C): pass
+class E(D): pass
+reveal_type(C.f)  # E: Revealed type is 'def () -> builtins.str*'
+reveal_type(C.f())  # E: Revealed type is 'builtins.str*'
+reveal_type(D.f)  # E: Revealed type is 'def () -> builtins.str*'
+reveal_type(D.f())  # E: Revealed type is 'builtins.str*'
+reveal_type(E.f)  # E: Revealed type is 'def () -> builtins.str*'
+reveal_type(E.f())  # E: Revealed type is 'builtins.str*'
+[builtins fixtures/classmethod.pyi]
+[out]
+
+[case testClassMethodOfMultipleGenericClass]
+from typing import TypeVar, Generic
+t = TypeVar('t')
+class A(Generic[t]):
+    @classmethod
+    def f(cls) -> t: ...
+class B(Generic[t]):
+    @classmethod
+    def f(cls) -> t: ...
+class C(A[str], B[int]): pass
+class D(A[int], B[str]): pass
+reveal_type(C.f)  # E: Revealed type is 'def () -> builtins.str'
+reveal_type(C.f())  # E: Revealed type is 'builtins.str'
+reveal_type(D.f)  # E: Revealed type is 'def () -> builtins.int'
+reveal_type(D.f())  # E: Revealed type is 'builtins.int'
+[builtins fixtures/classmethod.pyi]
+[out]
+
+[case testClassMethodOfGenericClassOverriding]
+from typing import Generic, TypeVar
+t = TypeVar('t')
+class A(Generic[t]):
+    @classmethod
+    def f(cls) -> t: pass
+class B(A[str]):
+    @classmethod
+    def f(cls) -> str: pass
+class C(A[str]):
+    @classmethod
+    def f(cls) -> int: pass  # E: Return type of "f" incompatible with supertype "A"
+[builtins fixtures/classmethod.pyi]
+[out]
+
+[case testClassMethodOfGenericClassWithBindingAnotherTypeVariable]
+from typing import Generic, TypeVar
+t = TypeVar('t')
+s = TypeVar('s')
+class A(Generic[t]):
+    @classmethod
+    def f(cls) -> t: pass
+class B(A[s]): pass
+reveal_type(A.f())  # E: Revealed type is '<nothing>'
+reveal_type(B.f())  # E: Revealed type is '<nothing>'
```

Comment:
What is this supposed to test? I would actually give an error if any variable stays unbound during the access via class object.

Reply:
I supposed I had wanted to check it works correctly when both replacing type variable T to S, and inheriting A from B.

But I didn't know clearly that how to represent. So I added like this ambiguously.

It may what I really wanted to represent like this:

```python
from typing import Generic, TypeVar
T = TypeVar('T')
S = TypeVar('S')
class A(Generic[T]):
    @classmethod
    def f(cls) -> T: pass
class B(A[S]): pass
reveal_type(A.f)  # E: Revealed type is 'def [T] () -> T`1'
reveal_type(B.f)  # E: Revealed type is 'def [S] () -> S`1'
```

Reply:
Yes, this test makes more sense.